2 단계 - 문자열 계산기

hyyg123 opened the pull request:
최소한의 요구사항을 만족하도록 개발하였습니다.

- StringCalculator 클래스를 통해 문자열 계산기를 구현하였습니다.
- kotest BehaviorSpec을 통해 테스트 코드를 작성하였습니다.
- Operator enum 클래스를 통해 연산자를 구현하였습니다.
- ErrorType enum 클래스를 통해 에러 메시지를 구현하였고, 예외 테스트 시에 활용하도록 했습니다.

(커밋  이력도 신경쓰려 했지만 빠른 학습을 위해 커밋 하나에 전부 구현하였습니다 하하)

한솔님 덕분에 개발 공부 자극이 된 것 같아 항상 고맙습니다 :)
리뷰 잘 부탁드려요 ~~

Commit:
docs : README 작성

- 2단계 문자열 계산기 요구사항 작성

Commit:
feat : 숫자 입력 더하기 기능 개발

- 숫자 입력 더하기 메소드 개발
- 숫자 입력 더하기 메소드 테스트 작성

hyyg123 committed:
chore : kotest version update

kotest version update
- 5.2.3 -> 5.4.0

hyyg123 committed:
feat : 문자열 계산기 기능 구현

문자열 계산기 구현
- StringCalculator 클래스 작성
- 문자열 계산기 구현
- OperatorTest, StringCalculatorTest 테스트 코드 작성

hyyg123 committed:
refactor : 에러 메시지 테스트 수정

에러 메시지 수정
- 에러 메시지 enum 클래스로 선언(ErrorType)
- 예외 테스트 시, 메시지 비교 로직 추가
- 빈 문자열 예외 처리

hyyg123 committed:
refactor : 소수점 계산 가능하도록 수정

소수점 계산
- Int -> Double 변경

lhs0610 commented on README.md:
```diff
@@ -1 +1,28 @@
+# kotlin-racingcar
+
+
+## 2단계 - 문자열 계산기
+### 기능 요구 사항
+
+- 사용자가 입력한 문자열 값에 따라 사칙 연산을 수행할 수 있는 계산기를 구현해야 한다.
+- 문자열 계산기는 사칙 연산의 계산 우선순위가 아닌 입력 값에 따라 계산 순서가 결정된다. 즉, 수학에서는 곱셈, 나눗셈이 덧셈, 뺄셈 보다 먼저 계산해야 하지만 이를 무시한다.
+- 예를 들어 "2 + 3 * 4 / 2"와 같은 문자열을 입력할 경우 2 + 3 * 4 / 2 실행 결과인 10을 출력해야 한다.
+
+### 프로그래밍 요구 사항
+
+- 메서드가 너무 많은 일을 하지 않도록 분리하기 위해 노력해 본다.
+- kotest를 이용해 구현해보도록 한다.
+
+### 힌트
+
+테스트할 수 있는 단위로 나누어 구현 목록을 만든다.
+
+- 덧셈
+- 뺄셈
+- 곱셈
+- 나눗셈
+- 입력값이 null이거나 빈 공백 문자일 경우 IllegalArgumentException throw
+- 사칙연산 기호가 아닌 경우 IllegalArgumentException throw
+- 사칙 연산을 모두 포함하는 기능 구현
+
+공백 문자열을 빈 공백 문자로 분리하려면 String 클래스의 split(" ") 메소드를 활용한다. 반복적인 패턴을 찾아 반복문으로 구현한다.
```
체크리스트 👍 

기능 요구사항에서 내용이 디테일해 질 수록 테스트 코드 작성에 용이성이 생기니 이 부분도 신경써보면 좋을 것 같아요 ㅎㅎ

lhs0610 commented on src/main/kotlin/stringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,36 @@
+import stringcalculator.constants.ErrorType.INVALID_INPUT_BLANK
+import stringcalculator.constants.ErrorType.INVALID_INPUT_NULL
+import stringcalculator.enums.Operator
+import java.util.*
```
import wildcard(*)는 지양해주는게 좋아요 소스파일 최상단의 패키지 목록을보면서 해당 소스 코드를 파악하기도 하는데, 그러한 가독성을 낮추기도해서 보통 IDE설정에서 변경해주는편입니다

lhs0610 commented on src/main/kotlin/stringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,36 @@
+
+class StringCalculator {
+    fun calculate(expression: String?): Double {
+        val splits = expression?.split(DELIMITER)
+            ?: throw IllegalArgumentException(INVALID_INPUT_NULL.message)
```
Elvis 연산자를 사용하셨네요 👍 
그런데, Null이 아닌 다른 기타 값들을 제공하거나 하위로직을 단순히 실행 안시키고 반환할게 아니라면, 
예외 제한 블럭을 통해 유효성 검증을 하는 방식으로 하면 어떨까요? 

해당 방식들을 통해 스마트캐스팅 기능도 제공이 되기 때문에 좀 더 가독성과 편이성을 높힐 수 있어요.
```suggestion
        require(!expression.isNullOrBlank()){ 
            INVALID_INPUT_NULL.message
        }
        
        //이 이후로 expression은 not null임이 스마트 캐스트 되어 보장됩니다.
```

lhs0610 commented on src/main/kotlin/stringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,36 @@
+        val splits = expression?.split(DELIMITER)
+            ?: throw IllegalArgumentException(INVALID_INPUT_NULL.message)
+
+        validateEmpty(expression)
```
위에서isNullOrBlank()로 null과 blank 테스트가 같이 진행되었다면 해당 검증은 필요 없겠죠

lhs0610 commented on src/main/kotlin/stringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,36 @@
+
+        validateEmpty(expression)
+
+        val queue: Queue<String> = LinkedList(splits)
```
LinkedList도 좋지만, Kotlin에서는 ArrayDeque라는 자료구조를 써보는건 어떨까요?
이를 권장하는 이유로 첫째는, kotlin스럽게라는 말에는 java의 라이브러리 import하는 것을 최소화 하자는 것이고, 둘 째로는 ArrayDeque가 LinkedList에 비해 성능 측면에서도 더 뛰어납니다([참고](https://www.baeldung.com/kotlin/arraydeque))

hyyg123 replied:
Kotlin에서는 java 라이브러리는 최소화해야겠군요!

lhs0610 commented on src/main/kotlin/stringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,36 @@
+        validateEmpty(expression)
+
+        val queue: Queue<String> = LinkedList(splits)
+        var result = queue.poll().toDouble()
+        while (queue.isNotEmpty()) {
+            val operator = Operator.of(queue.poll())
+            val operand = queue.poll().toDouble()
+
+            result = operator.calculate(result, operand)
+        }
```
코틀린의 컬렉션 API를 활용해보면 어떨까요?

lhs0610 commented on src/main/kotlin/stringcalculator/StringCalculator.kt:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    companion object {
+        private const val DELIMITER = " "
```
매직리터럴 분리 👍

lhs0610 commented on src/main/kotlin/stringcalculator/enums/Operator.kt:
```diff
@@ -0,0 +1,23 @@
+    fun calculate(num1: Double, num2: Double): Double = operate(num1, num2)
+
+    companion object {
+        fun of(input: String): Operator {
+            return values().firstOrNull { operator -> operator.symbol == input }
+                ?: throw IllegalArgumentException(ErrorType.INVALID_OPERATOR.message)
+        }
```
매번 values()를 호출해서 찾는것보단 값들을 적절한 키(ex: symbol)로 캐싱해서 찾도록 하면 어떨까요?
추가적으로 values()는 매 번 새로운 배열을 생성하기 때문에 [성능 이슈](https://github.com/Kotlin/KEEP/blob/master/proposals/enum-entries.md?fbclid=IwAR0RUuWmASChIRcDUxdvPa-MrqaM2hoAIjT16tKLEepbWdBPpB6oKz-Qjro&mibextid=S66gvF#examples-of-performance-issues)가 있을 수 있습니다.(kotlin 1.8.2x 버전부터는 상관없다고 합니다.)

lhs0610 commented on src/test/kotlin/stringcalculator/OperatorTest.kt:
```diff
@@ -0,0 +1,21 @@
+import stringcalculator.constants.ErrorType.INVALID_OPERATOR
+import stringcalculator.enums.Operator
+
+class OperatorTest : BehaviorSpec({
```
정상적인 연산자를 제공했을때 의도하는 연산자가 나오는지도 있어야 할 것 같네요 ㅎㅎ

lhs0610 commented on src/test/kotlin/stringcalculator/OperatorTest.kt:
```diff
@@ -0,0 +1,21 @@
+
+        When("잘못된 연산자로 Operator 를 찾을 때") {
+            Then("IllegalArgumentException 예외를 던진다.") {
+                val exception = shouldThrow<IllegalArgumentException> { Operator.of(input) }
```
실패 테스트 👍

lhs0610 commented on src/test/kotlin/stringcalculator/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,160 @@
+import io.kotest.matchers.string.startWith
+import stringcalculator.constants.ErrorType.*
+
+class StringCalculatorTest : BehaviorSpec({
```
값을 0으로 나누려할때나 0을 값으로 나누려할때에 대한 테스트도 필요할 것 같네요!

lhs0610 reviewed (changes requested):
용구님 2단계 문자열 계산기 미션 밤늦게 고생하셨습니다 👍 
기능적으로 문제는 없으나 코틀린스럽게 좀 더 리팩토링하면 좋을 부분들에 대해 피드백 남겨드렸으니 확인 후 다시 리뷰요청 부탁드려요

Commit:
refactor : PR 피드백 반영

- 상세 요구사항 추가
- LinkedList -> ArrayDeque 변경
- Operator of 메소드 캐싱
- Operator DIVIDE 0으로 나눌 때 예외 처리 (테스트 추가)
- Operator 정상 테스트 추가
- StringCalculator class 검증 로직 리팩토링

hyyg123 commented:
@lhs0610 

피드백 반영하였습니다.

반영사항
- 상세 요구사항 추가
- LinkedList -> ArrayDeque 변경
- Operator of 메소드 캐싱
- Operator DIVIDE 0으로 나눌 때 예외 처리 (테스트 추가)
- Operator 정상 테스트 추가 
- StringCalculator 검증 로직 리팩토링

확인 부탁드려요 ~

lhs0610 commented on src/main/kotlin/stringcalculator/StringCalculator.kt:
```diff
@@ -1,35 +1,28 @@
+        }
 
-        validateEmpty(expression)
+        val splits = expression.split(DELIMITER)
```
해당 구분자가 아닌 다른 로직을 통해 분리될 수도 있도록 확장성을 고려해도 좋을 것 같아요.
기획 변경이 있을때마다 해당 로직을 수정하면 OCP원칙에도 위배될 수 있겠죠. 추상화를 이용해 전략 패턴을 활용해봐도 좋을 것 같아요.

hyyg123 replied:
네, 해당 사항 3단계에서 피드백 반영하도록 할게요!

lhs0610 reviewed (approved):
용구님 빠른 피드백 반영 감사합니다 👍 

전체적으로 코틀린 기능도 잘 활용하셨고, 테스트도 꼼꼼하게 작성되신 것 같아요 💯 
다만, 다음 미션부터는 좀 더 도메인이나 계층 분리를 활용해서 SRP 원칙을 고민해보면 좋을 것 같습니다. 계산기 미션은 여기서 merge하도록 하겠습니다.
